### PR TITLE
Trying to prevent accidental usage of the fuzztarget functions 

### PR DESCRIPTION
**Pull request by elichai**
A suggestion of a way to solve https://github.com/rust-bitcoin/rust-secp256k1/issues/90

This is a result of a conversation with @real-or-random and some feedback from @TheBlueMatt 

The part i'm less sure of is the tests here, because rust tests run in concurrent threads and I want to test both with the bool set and without I had to use some Mutex to make sure they don't run together

Another thing that isn't too nice is that we need to put this is all the fuzzing functions.
if proc macros were stabilized before 1.22 we could've wrote an attribute macro and made it a bit nicer but it would've still required doing it on every function.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1084,3 +1084,84 @@ mod benches {
...
+        INIT_CONTEXT.call_once(|| unsafe {
+            CRYPTO_VAR_LOCK = Some(Default::default());
+        });
+        unsafe { CRYPTO_VAR_LOCK.as_ref().unwrap().lock().unwrap() }
+    }
+
+
+
+    pub fn crashed_for_fuzz<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> bool {
+        let description = "Tried fuzzing without setting the UNSAFE_CRYPTO_FUZZING variable";
```
Comment (real-or-random) [inline src/lib.rs]: Maybe reduce this string just to "UNSAFE_CRYPTO_FUZZING", that's enough for a test.

File: src/ffi.rs [hunk 2]
```diff
@@ -408,36 +411,43 @@ mod fuzz_dummy {
     // Contexts
     /// Creates a dummy context, tracking flags to ensure proper calling semantics
     pub unsafe fn secp256k1_context_preallocated_create(_ptr: *mut c_void, flags: c_uint) -> *mut Context {
+        assert!(UNSAFE_CRYPTO_FUZZING, "Tried fuzzing without setting the UNSAFE_CRYPTO_FUZZING variable");
```
Comment (real-or-random) [inline src/ffi.rs]: A constant for the string seems to be a good idea. Maybe even a function / macro `assert_fuzzing()`. (Well it's a one-liner, so it does not make much of a difference if it's a function call or the assert directly. I think both ways are fine.)

**Comment by elichai** [comment]
> Well it's not super elegant but it works, so it's good. This is discussed here: [rust-lang/rust#43155](https://github.com/rust-lang/rust/issues/43155) This issue has some nice code snippets. The simplest solution is to pass `-- --test-threads=1` to `cargo test`. I think we could actually do that (with a comment that explains it).

That's cool :) did not know about this flag.

File: .travis.yml [hunk 3]
```diff
@@ -32,6 +32,7 @@ script:
   - cargo test --verbose --features=rand
   - cargo test --verbose --features="rand serde"
   - cargo test --verbose --features="rand serde recovery endomorphism"
+  - cargo test --lib --features=fuzztarget -- --test-threads=1 # 1 thread is required so that we can have 2 tests setting the `UNSAFE_CRYPTO_FUZZING` variable.
```
Comment (real-or-random) [inline .travis.yml]: Shouldn't this be `--verbose` instead of `--lib`?

Comment (elichai) [reply .travis.yml]: the `--lib` is to prevent the docs tests from running, because they don't pass.

I'll also add the --verbose

**Comment by elichai** [comment]
Rebased, and replaced the `-- --test-threads=1` trick with just running them both in the same single test.

**Comment by apoelstra** [comment]
Closing, fixed by #263

**Pull request closed without merging**